### PR TITLE
Pygeoapi provider

### PR DESCRIPTION
**Pull request by faysal-ishtiaq**
First attempt at creating a CDMS Provider for [pygeoapi](https://pygeoapi.io/).

The long term aim is to allow `from opencdms.pygeoapi import CDMSProvider` which can be configured to work with any supported CDMS.

The current version only implements [ClimsoftProvider](https://github.com/opencdms/pyopencdms/pull/47/files#diff-0b8a69aaf7ddcf393c5920523153c809acc5259dcc602d477b7c995b49bdf811R105) for working with Climsoft CDMS:
`from opencdms.pygeoapi.climsoft_provider import ClimsoftProvider`.

- [x] publish `observationfinal` table as a collection
- [x] visualize on `items` page
- [x] visualize single `item`
- [x] implement sorting
- [x] test sorting
- [x] implement filtering
- [x] test filtering
- [x] implement bbox filtering
- [x] test bbox filtering

### CODE REVIEW

**Comment by faysal-ishtiaq** [comment]
@isedwards 

I have added this last section in main readme to explain how to use climsoft provider: https://github.com/opencdms/pyopencdms/tree/pygeoapi-provider#running-pygeoapi-server

I have updated this readme to demonstrate the current state of climsoft provider
https://github.com/opencdms/pyopencdms/tree/pygeoapi-provider/opencdms/pygeoapi

**Comment by faysal-ishtiaq** [comment]
@isedwards 


In this image, the bbox is [-180, -90, 180, 90] - which covers the whole map and we get all the data

Screenshot from 2022-09-28 21-39-02

In this image the bbox is [1, 1, 180, 90] - this returns no data. Because our station is located at latitude: -17.83
longitude: 31.02 and there is no station with all positive lat/lng

Screenshot from 2022-09-28 21-39-14

**Comment by isedwards** [comment]
@faysal-ishtiaq - this is looking good.  Are the id's all the same at the moment (are we return the station id for every result)?

Any thoughts on why we're not seeing the background map?

**Comment by faysal-ishtiaq** [comment]
> @faysal-ishtiaq - this is looking good. Are the id's all the same at the moment (are we return the station id for every result)?
> 
> Any thoughts on why we're not seeing the background map?

We have observation data from only one station. And apparently that station is in the sea

image

**Comment by isedwards** [comment]
Are the lon (x) and lat (y) the right way around?

**Comment by faysal-ishtiaq** [comment]
> Are the lon (x) and lat (y) the right way around?

here it says that longitude is x and latitude is y

https://gis.stackexchange.com/questions/11626/does-y-mean-latitude-and-x-mean-longitude-in-every-gis-software#:~:text=Latitude%20is%20the%20Y%20axis,180%20longitude%20system%20is%20use.

**Comment by isedwards** [comment]
Agreed, but it looks like you've plotted (lat, lon) instead of (lon, lat). Could you try the coordinates the other way round? If you do, it looks likely that the station will be plotted correctly in Zimbabwe.

**Comment by faysal-ishtiaq** [comment]
yes, it was an error from my side. I have fixed it

image

File: opencdms/pygeoapi/climsoft_provider.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+        Get provider field information (names, types)
+
+        :returns: dict of fields
+        """
+        if not self.fields:
+            with DatabaseConnection(
+                self.conn_dic, properties=self.properties
+            ) as db:
+                self.fields = db.fields
+        return self.fields
```
Comment (faysal-ishtiaq) [inline opencdms/pygeoapi/climsoft_provider.py:166]: I checked this piece of code. It returns an OrderedDict

```
OrderedDict([('recordedFrom', {'type': 'string', 'maxLength': 255}), ('describedBy', {'type': 'string'}), ('obsDatetime', {'type': 'string', 'format': 'date-time'}), ('obsLevel', {'type': 'string', 'maxLength': 255}), ('obsValue', {'type': 'integer'}), ('flag', {'type': 'string', 'maxLength': 255}), ('period', {'type': 'integer'}), ('qcStatus', {'type': 'integer'}), ('qcTypeLog', {'type': 'string'}), ('acquisitionType', {'type': 'integer'}), ('dataForm', {'type': 'string', 'maxLength': 255}), ('capturedBy', {'type': 'string', 'maxLength': 255}), ('mark', {'type': 'integer'}), ('temperatureUnits', {'type': 'string', 'maxLength': 255}), ('precipitationUnits', {'type': 'string', 'maxLength': 255}), ('cloudHeightUnits', {'type': 'string', 'maxLength': 255}), ('visUnits', {'type': 'string', 'maxLength': 255}), ('dataSourceTimeZone', {'type': 'integer'})])
```

File: opencdms/pygeoapi/climsoft_provider.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+        if bbox is None:
+            bbox = []
+        if properties is None:
+            properties = []
+        if sortby is None:
+            sortby = []
+        if select_properties is None:
+            select_properties = []
+        LOGGER.error(sortby)
```
Comment (faysal-ishtiaq) [inline opencdms/pygeoapi/climsoft_provider.py]: But I don't get anything here. It returns an empty list 

```bash
[2022-10-16T11:21:18Z] {/home/faysal/PycharmProjects/pyopencdms/venv/lib/python3.8/site-packages/opencdms/pygeoapi/climsoft_provider.py:192} ERROR - []
```

File: opencdms/pygeoapi/climsoft_provider.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+                .filter(models.Station.longitude >= min_lng)
+                .filter(models.Station.longitude <= max_lng)
+                .filter(models.Station.latitude >= min_lat)
+                .filter(models.Station.latitude <= max_lat)
+            )
+
+        return query
+
+    def _apply_sorting(self, query: Query, sortby: List[Dict]):
+        LOGGER.error(sortby)
```
Comment (faysal-ishtiaq) [inline opencdms/pygeoapi/climsoft_provider.py]: enpty list here as well

**Comment by faysal-ishtiaq** [comment]
Also for some url (such as, http://localhost:5000/collections/climsoft/items?obsDatetime=2000-01-19%2006:00:00&sortby=+obsValue ), I get the following error

image


When I looked into for the error message, I found it here https://github.com/geopython/pygeoapi/blob/master/pygeoapi/api.py#L1419

Which implies that the sort key was not found in `db.fields` . But `obsValue` is in the fields (the OrderedDict)

```
>>> "obsValue" in a.keys()
True
```

I am not sure what I am missing.

**Pull request merged**